File: server/csharp/MetaObjects.Codegen/TemplateCodegen/TemplateSpec.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using MetaObjects.Codegen.Generators;
using MetaObjects.Render;

namespace MetaObjects.Codegen.TemplateCodegen;

/// <summary>One declarative template-generator spec entry (SP-1 §4 cross-port contract).</summary>
public sealed record TemplateSpecEntry(
    string Name, string Template, string Scope, string OutputPattern, string? Format);

/// <summary>
/// The declarative JSON template-spec the CLI ports (C#/Python) consume. The JSON
/// shape is the cross-port contract; a JSON Schema sits beside the TS port. This
/// class validates it and maps it to runnable <see cref="IGenerator"/>s.
/// </summary>
public static class TemplateSpec
{
    private static readonly HashSet<string> ValidFormats = new(StringComparer.Ordinal)
    {
        Escapers.FORMAT_TEXT, Escapers.FORMAT_HTML, Escapers.FORMAT_XML, Escapers.FORMAT_CSV,
        Escapers.FORMAT_JSON, Escapers.FORMAT_MARKDOWN, Escapers.FORMAT_SPREADSHEET,
    };

    private static string ReqStr(JsonElement entry, int i, string key)
    {
        if (!entry.TryGetProperty(key, out var v) || v.ValueKind != JsonValueKind.String
            || v.GetString() is not { Length: > 0 } s)
            throw new ArgumentException($"template-spec generators[{i}]: missing or empty required string '{key}'");
        return s;
    }

    public static IReadOnlyList<TemplateSpecEntry> Parse(JsonElement root)
    {
        if (root.ValueKind != JsonValueKind.Object
            || !root.TryGetProperty("generators", out var gens)
            || gens.ValueKind != JsonValueKind.Array)
            throw new ArgumentException("template-spec: expected an object with a `generators` array");

        var entries = new List<TemplateSpecEntry>();
        var i = 0;
        foreach (var entry in gens.EnumerateArray())
        {
            if (entry.ValueKind != JsonValueKind.Object)
                throw new ArgumentException($"template-spec generators[{i}]: expected an object");
            var name = ReqStr(entry, i, "name");
            var template = ReqStr(entry, i, "template");
            var scope = ReqStr(entry, i, "scope");
            var outputPattern = ReqStr(entry, i, "outputPattern");
            if (!ScopeWalk.Scopes.Contains(scope))
                throw new ArgumentException(
                    $"template-spec generators[{i}]: scope must be one of {string.Join(" | ", ScopeWalk.Scopes)}, got '{scope}'");

            string? format = null;
            if (entry.TryGetProperty("format", out var f) && f.ValueKind == JsonValueKind.String)
            {
                format = f.GetString();
                if (!ValidFormats.Contains(format!))
                    throw new ArgumentException(
                        $"template-spec generators[{i}]: format must be one of {string.Join(" | ", ValidFormats)}, got '{format}'");
            }

            // `target` (output routing) is not supported by the C# port — EmittedFile.Path is
            // relative to a single out dir. Reject it loudly so the cross-port divergence is
            // explicit rather than a silent layout difference vs TS.
            if (entry.TryGetProperty("target", out _))
                throw new ArgumentException(
                    $"template-spec generators[{i}]: `target` is not supported by the C# port (it has no output-target concept; output is relative to a single out dir)");

            entries.Add(new TemplateSpecEntry(name, template, scope, outputPattern, format));
            i++;
        }
        return entries;
    }

    public static IReadOnlyList<IGenerator> ToGenerators(IReadOnlyList<TemplateSpecEntry> spec, IProvider provider) =>
        spec.Select(e => TemplateGenerator.Create(
            e.Name, e.Template, ScopeWalk.ForScope(e.Scope, e.OutputPattern), provider, e.Format ?? Escapers.FORMAT_TEXT))
            .ToList();
}

File: server/python/src/metaobjects/codegen/template_codegen/template_spec.py
"""The declarative JSON template-spec the CLI ports (Python/C#) consume.

The JSON shape is the cross-port contract (SP-1 §4) — a JSON Schema sits beside
the TS port at codegen-ts/src/template-codegen/template-spec.schema.json. This
module validates + maps it to runnable Generators.
"""

from __future__ import annotations

from typing import Any

from metaobjects.codegen.generator import Generator
from metaobjects.codegen.generators.template_generator import SCOPES, template_generator
from metaobjects.render import escapers
from metaobjects.render.verify import Provider

_FORMATS = (
    escapers.FORMAT_TEXT,
    escapers.FORMAT_HTML,
    escapers.FORMAT_XML,
    escapers.FORMAT_CSV,
    escapers.FORMAT_JSON,
    escapers.FORMAT_MARKDOWN,
    escapers.FORMAT_SPREADSHEET,
)

_REQUIRED_STR = ("name", "template", "scope", "outputPattern")


def parse_template_spec(obj: object) -> dict[str, Any]:
    """Validate + return a normalized ``{"generators": [...]}`` dict. Raises
    ``ValueError`` on any shape violation."""
    if not isinstance(obj, dict) or not isinstance(obj.get("generators"), list):
        raise ValueError("template-spec: expected an object with a `generators` array")
    generators: list[dict[str, Any]] = []
    for i, raw in enumerate(obj["generators"]):
        if not isinstance(raw, dict):
            raise ValueError(f"template-spec generators[{i}]: expected an object")
        for key in _REQUIRED_STR:
            if not isinstance(raw.get(key), str) or raw[key] == "":
                raise ValueError(
                    f"template-spec generators[{i}]: missing or empty required string {key!r}"
                )
        if raw["scope"] not in SCOPES:
            raise ValueError(
                f"template-spec generators[{i}]: scope must be one of "
                f"{' | '.join(SCOPES)}, got {raw['scope']!r}"
            )
        entry: dict[str, Any] = {
            "name": raw["name"],
            "template": raw["template"],
            "scope": raw["scope"],
            "outputPattern": raw["outputPattern"],
        }
        if "format" in raw:
            if raw["format"] not in _FORMATS:
                raise ValueError(
                    f"template-spec generators[{i}]: format must be one of "
                    f"{' | '.join(_FORMATS)}, got {raw['format']!r}"
                )
            entry["format"] = raw["format"]
        if isinstance(raw.get("target"), str):
            entry["target"] = raw["target"]
        generators.append(entry)
    return {"generators": generators}


def template_spec_to_generators(spec: dict[str, Any], provider: Provider) -> list[Generator]:
    """Map a parsed spec into runnable Generators (one per entry). The caller
    supplies the ``provider`` (the CLI builds a ``FilesystemProvider``)."""
    out: list[Generator] = []
    for e in spec["generators"]:
        out.append(
            template_generator(
                name=e["name"],
                template=e["template"],
                scope=e["scope"],
                output_pattern=e["outputPattern"],
                provider=provider,
                format=e.get("format", escapers.FORMAT_TEXT),
            )
        )
    return out
